scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/AsyncFlowDefRunner.scala

```diff
@@ -0,0 +1,201 @@
+        }
+      }
+    }
+  }
+}
+
+/**
+ * This holds an internal thread to submit run
+ * a Config, Mode, FlowDef and return a Future holding the
+ * JobStats
+ */
+class AsyncFlowDefRunner extends Writer { self =>
+  import AsyncFlowDefRunner._
+
+  private[this] val filesToCleanup = mutable.Set[String]()
+  private val messageQueue: LinkedBlockingQueue[AsyncFlowDefRunner.FlowDefAction] =
+    new LinkedBlockingQueue[AsyncFlowDefRunner.FlowDefAction]()
+
+  /**
+   * Hadoop and/or cascading has some issues, it seems, with starting jobs
+   * from multiple threads. This thread does all the Flow starting.
+   */
+  private lazy val thread = new Thread(new Runnable {
+    def run(): Unit = {
+      @annotation.tailrec
+      def go(id: Long): Unit = messageQueue.take match {
+        case Stop => ()
+        case RunFlowDef(conf, mode, fd, promise) =>
+          try {
+            val ctx = ExecutionContext.newContext(conf)(fd, mode)
+            ctx.buildFlow match {
+              case Success(flow) =>
+                val future = FlowListenerPromise
+                  .start(flow, { f: Flow[_] => (id, JobStats(f.getFlowStats)) })
+
+                promise.completeWith(future)
+              case Failure(err) =>
+                promise.failure(err)
+            }
+          } catch {
+            case t: Throwable =>
+              // something bad happened, but this thread is a daemon
+              // that should only stop if all others have stopped or
+              // we have received the stop message.
+              // Stopping this thread prematurely can deadlock
+              // futures from the promise we have.
+              // In a sense, this thread does not exist logically and
+              // must forward all exceptions to threads that requested
+              // this work be started.
+              promise.tryFailure(t)
+          }
+          // Loop
+          go(id + 1)
+      }
+
+      // Now we actually run the recursive loop
+      go(0)
+    }
+  })
+
+  def runFlowDef(conf: Config, mode: Mode, fd: FlowDef): Future[(Long, JobStats)] =
+    try {
+      val promise = Promise[(Long, JobStats)]()
+      val fut = promise.future
+      messageQueue.put(RunFlowDef(conf, mode, fd, promise))
+      // Don't do any work after the .put call, we want no chance for exception
+      // after the put
+      fut
+    } catch {
+      case NonFatal(e) =>
+        Future.failed(e)
+    }
+
+  def start(): Unit = {
+    // Make sure this thread can't keep us running if all others are gone
+    thread.setDaemon(true)
+    thread.start()
+  }
+  /*
+   * This is called after we are done submitting all jobs
+   */
+  def finished(mode: Mode): Unit = {
+    messageQueue.put(Stop)
+    // get an immutable copy
+    val cleanUp = filesToCleanup.synchronized { filesToCleanup.toList }
+    if (cleanUp.nonEmpty) {
+      Runtime.getRuntime.addShutdownHook(TempFileCleanup(cleanUp, mode))
```

Comment:
I wasn't that clear: if we do the above, the logic about cleanup files can happen when you do forceToDisk, which Writer is driving, so it can keep the list of clean-up files it needs.